Allow to pass a custom OrgEnv instance when loading content

schlagenhauf opened the pull request:
In the current implementation the OrgEnv instance is created inside the
`parse_lines()` method with the only customizable argument being the
filename. In this change the methods `load()`/`loads()`/`loadi()` are extended by an optional `env` argument to pass it in when loading content.

Fixes #18

karlicoss commented on orgparse/node.py:
```diff
@@ -10,6 +10,18 @@
 from .utils.py3compat import PY3, unicode
 
 
+def parse_lines(lines, env):
```
any reason to move it to the top of the file? IMO generally better to keep the code as intact as possible to make `git blame` easier

schlagenhauf replied:
I moved it there while diving into the architecture. Keeping changes minimal is a valid point though, i'll change it.

karlicoss commented on orgparse/__init__.py:
```diff
@@ -138,25 +139,27 @@ def load(path):
 
 
-def loadi(lines, filename='<lines>'):
+def loadi(lines, env=OrgEnv(filename='<lines>')):
```
Might be unsafe because OrgEnv object is mutable (and it will be kind of like global variable).
The way you did in [lines 143-144](https://github.com/karlicoss/orgparse/pull/19/files#diff-279135a213a417a10e945ccc26dfeadeR143-R144) (i.e. set to None and assign inside the function) is good!

schlagenhauf replied:
True! Will fix.

karlicoss commented on orgparse/__init__.py:
```diff
@@ -138,25 +139,27 @@ def load(path):
-                 filename=filename)
 
+    if not env:
+        env = OrgEnv(filename=filename)
```
I had in mind doing this bit in `parse_lines` (so it would have signature `def parse_lines(lines, filename, env=None)`. That way it would be handled in one place and will be backwards compatible (someone might have been using `parse_lines`) before. What do you think?

karlicoss replied:
P.S. also meant that `loadi`/`loads` etc would also have extra `env=None` argument that they will simply pass down as `env=env`

schlagenhauf replied:
I tested a few signatures but wasn't convinced. In your example, something like `parse_lines(my_lines, "filename1", OrgEnv(filename="filename2"))` may lead to unexpected behavior.

But i totally agree, it's not a good idea to break the API for this.

karlicoss replied:
Good point too about filenames, but hopefully ignoring `filename` argument is not a big problem if the user is passing `OrgEnv` down (presumably they know what they are doing). Also when there is no proper filename it's `<undefined>/<lines>/<string>` anyway, so a placeholder only used for debugging.

karlicoss reviewed (changes requested):
Thanks, have few comments!

schlagenhauf committed:
Allow to pass a custom OrgEnv instance when loading content

In the previous version the OrgEnv instance was created inside the
parse_lines() method with the only customizable argument being the
filename. In this change the creation of the instance happens in the
load/loads/loadi methods and can be optionally passed as an argument
from outside.

schlagenhauf committed:
Add unit tests for setting custom todo keys

schlagenhauf committed:
Remove unneccesary conditional in argument check

karlicoss commented on orgparse/tests/test_misc.py:
```diff
@@ -33,3 +35,50 @@ def test_stars():
     assert h2.heading == ''
 
+
+def test_parse_custom_todo_keys():
```
awesome, thanks for detailed tests!

karlicoss commented:
Sweet, looks great, thanks!

I'll release it on pypi later tonight/tomorrow morning!

karlicoss commented:
Done! Feels good to finally automate releases from CI based on git tags.

schlagenhauf commented:
Awesome, thank you for your super fast response time! Just noticed the CI additions, great stuff!
